typing/includecore.mli

```diff
@@ -20,20 +20,48 @@ open Types
 
 exception Dont_match
 
+type label_mismatch =
+  | Type
+  | Mutable of bool
+
+type record_mismatch =
+  | Label_mismatch of label_declaration * label_declaration * label_mismatch
+  | Label_names of int * Ident.t * Ident.t
+  | Label_missing of bool * Ident.t
+  | Representation of bool   (* true means second one is unboxed float *)
+
+type constructor_mismatch =
+  | Type
+  | Arity
+  | Record of record_mismatch
+  | Kind of bool
+  | Explicit_return_type of bool
+
+type variant_mismatch =
+  | Constructor_mismatch of constructor_declaration
+                            * constructor_declaration
+                            * constructor_mismatch
+  | Constructor_names of int * Ident.t * Ident.t
+  | Constructor_missing of bool * Ident.t
+
+type extension_constructor_mismatch =
+  | Privacy
+  | Constructor_mismatch of Ident.t
+                            * extension_constructor
+                            * extension_constructor
+                            * constructor_mismatch
+
 type type_mismatch =
-    Arity
+  | Arity
   | Privacy
   | Kind
   | Constraint
   | Manifest
   | Variance
-  | Field_type of Ident.t
-  | Field_mutable of Ident.t
-  | Field_arity of Ident.t
-  | Field_names of int * Ident.t * Ident.t
-  | Field_missing of bool * Ident.t
-  | Record_representation of bool
-  | Unboxed_representation of bool
+  | Record_mismatch of record_mismatch
+  | Variant_mismatch of variant_mismatch
+  | Extension_constructor_mismatch of extension_constructor_mismatch
```

Comment:
API design question: why do the `Label_mismatch` and `Constructor_mismatch` constructors carry the conflicting declarations, but the `*_mismatch` constructors here (for records, vairants and extension constructors) don't? Do we need the information only in one case and not the other? Or is it just that is is easy to recompute it here, but harder in the case of labels/fields? (Wouldn't it be nicer to be consistent in the approach?)

Reply:
I didn't need to print whole records and variants but just constructors and fields to specify error. Also the type is specified by single record or variant declaration, which is already kept in Includemod.symptom type withing Type_declarations constructor.